### DISCUSSION TITLE
Visualising attention

### DISCUSSION THREAD

**Discussion by ilyalasy** [facebookresearch/ParlAI#4424] [Q&A]
Hey, I was wondering why ParlAI has no framework for outputting attention weights from models for later visualisation?
I think this is pretty important for interpretability of large language models. Or I have missed something?

For example, Huggingface has [output_attentions](https://huggingface.co/docs/transformers/main_classes/output#:~:text=output_hidden_states%3DTrue%20or-,output_attentions%3DTrue,-.) parameter which can help to use such libraries as [bertviz](https://github.com/jessevig/bertviz), [lit](https://github.com/PAIR-code/lit), etc.

**Comment by klshuster** [comment]
Both our [`BasicAttention`](https://github.com/facebookresearch/ParlAI/blob/4f7b38e7970424e4329cb57ab65710291a50f3f7/parlai/agents/transformer/modules/attention.py#L83-L84) and [`MultiheadAttention`](https://github.com/facebookresearch/ParlAI/blob/4f7b38e7970424e4329cb57ab65710291a50f3f7/parlai/agents/transformer/modules/attention.py#L252-L268) weights can (and do) output the attention weights from their forward pass (for `MultiheadAttention`, these are pre-softmax), though they are indeed ignored in the higher levels of the model (at least for our `transformer/generator` models)

**Reply by ilyalasy** [reply]
Yeah, I looked through the code and the output is non retrievable, I am quickly writing some clunky code for retrieving attention weights and tokens, maybe will post it here if interested

Are you interested in adding this feature later, maybe we should create an issue?

**Reply by stephenroller** [reply]
If you look at the distillation code, we use hooks to actually pull out the attention weights. You could use the same trick elsewhere. This code could be simplified but you still get the idea:

https://github.com/facebookresearch/ParlAI/blob/f773ef7281907401fcb8b081e0639b99f8a1cd94/projects/anti_scaling/distillation.py#L256-L272

**Reply by ilyalasy** [reply]
This is very helpful, thank you!
